### PR TITLE
Set up caching

### PR DESCRIPTION
**Pull request by tsauerwein**
See https://github.com/c2corg/v6_api/wiki/Caching-strategy#caching-for-the-api-project

* [x] Add a table `cache_versions` which has a version number for each document that reflects the state of the document and its associations.
* [x] Update cache key when a document or its associations change.
* [ ] Update cache key when map or area changes.
* [ ] Check that cache keys are updated when user name is changed (outings, document versions). 
* [x] Set up dogpile.cache for `/{type}/{document_id}` views
* [x] Set cache key as `ETAG` (using [etag_cache](http://docs.pylonsproject.org/projects/pylons-webframework/en/latest/modules/controllers_util.html#pylons.controllers.util.etag_cache))
* [x] Use cache for list views/advanced search
* [ ] Use cache for simple search
* [x] Use cache for history/version views
* [ ] Clear cache script

### CODE REVIEW

**Comment by tsauerwein** [comment]
Ready to review. I will create new issues for the remaining tasks before merging the PR.

File: c2corg_api/models/cache_version.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+    When the document or one of its associated documents is updated, the
+    version of the document is incremented, so that outdated documents are not
+    served from the cache.
+    """
+    __tablename__ = 'cache_versions'
+
+    document_id = Column(
+        Integer, ForeignKey(schema + '.documents.document_id'),
+        primary_key=True)
+    document = relationship(
```
Comment (asaunier) [inline c2corg_api/models/cache_version.py]: Missing blank space maybe :)

File: c2corg_api/models/cache_version.py [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+BEGIN
+  -- function to update all dependent documents if a document changes
+
+  -- update the version of the document itself
+  PERFORM guidebook.increment_cache_version(p_document_id);
+
+  -- update the version of linked documents (direct associations)
+  PERFORM guidebook.update_cache_version_of_linked_documents(p_document_id);
+
+  if p_document_type = 'w' then
```
Comment (asaunier) [inline c2corg_api/models/cache_version.py]: What about other document types to come (images, articles, profiles, books, accident reports)?
Will they required custom ``update_cache_version_*`` functions?

File: c2corg_api/models/cache_version.py [hunk 3]
```diff
@@ -295,3 +299,19 @@ def update_cache_version_associations(
...
+    version = DBSession.query(CacheVersion.version). \
+        filter(CacheVersion.document_id == document_id). \
+        first()
+
+    if not version:
+        # no version for this document id, the document should not exist
+        log.debug('no version for document id {0}'.format(document_id))
+        return
+
+    if not lang:
```
Comment (asaunier) [inline c2corg_api/models/cache_version.py]: In what cases will we cache documents without considering the lang?

File: c2corg_api/views/__init__.py [hunk 4]
```diff
@@ -157,3 +162,36 @@ def get_best_locale(available_locales, preferred_lang):
                 (available_locales[lang] for lang in langs_priority
                  if lang in available_locales), None)
     return best_locale
+
+
+def etag_cache(request, key):
+    """Use the HTTP Entity Tag cache for Browser side caching
+    If a "If-None-Match" header is found, and equivilant to ``key``,
```
Comment (asaunier) [inline c2corg_api/views/__init__.py]: "equivalent" maybe?

**Comment by asaunier** [comment]
This is huge! That looks really good to me :)

Since the system is quite complexe, that would probably be useful to have a README somewhere explaining how the cache system works.

I guess changes are required on the UI side to implement the etag approach. Do you plan to also have dogpile cache and etags (for the final end browsers) on the UI side?

File: c2corg_api/models/cache_version.py (hunk 2 above)
Comment (tsauerwein) [inline c2corg_api/models/cache_version.py]: If they have special dependencies like the ones below, yes.

**Comment by asaunier** [comment]
What happens if a document (nor its associated documents) does not change for a while? It stays cached forever, until a change is done?

What if the cache system (redis?) gets full? Older cache files are removed to save room for new cached documents?

Will it be possible to invalidate a cached document manually using some sort of request?
In v5, moderators have a button to clear a document cache. It is generally not needed except when using some actions that are not designed to invalidate the cache automatically. Perhaps we don't need such a feature in v6 if the cache invalidation is always done.

File: c2corg_api/models/cache_version.py (hunk 3 above)
Comment (tsauerwein) [inline c2corg_api/models/cache_version.py]: Getting a document with `/{type}/{id}` is a valid request for the API though I think it is not used in the UI.

**Comment by tsauerwein** [comment]
> Since the system is quite complex, that would probably be useful to have a README somewhere explaining how the cache system works.

I am planning to update the caching Wiki page with details.

> What happens if a document (nor its associated documents) does not change for a while? It stays cached forever, until a change is done?
> 
> What if the cache system (redis?) gets full? Older cache files are removed to save room for new cached documents?

All this is managed by Redis. Redis decides if and when a cached value is invalidated.

> Will it be possible to invalidate a cached document manually using some sort of request?

It would be easy to add an API endpoint which simply increments the `cache_key` of a document.

**Comment by asaunier** [comment]
What about a dev mode? Some way to disable the caching system when modifying the code of the API. If documents are cached, it may be difficult to test changes in the code?!

**Comment by tsauerwein** [comment]
> What about a dev mode? Some way to disable the caching system when modifying the code of the API. If documents are cached, it may be difficult to test changes in the code?!

Good point! I made it that a timestamp is appended to the cache key when in dev. mode. So that every time the dev. server re-loads, the cache key changes.

I also made the Git revision part of the cache key, so that the cached values are invalidated when the code changes.

Besides I am thinking about providing a "clear cache" script, which should be run when a new version of the API is deployed, which removes old cache values (basically it clears the Redis database).

**Comment by asaunier** [comment]
> Besides I am thinking about providing a "clear cache" script, which should be run when a new version of the API is deployed, which removes old cache values (basically it clears the Redis database).

That would make sense indeed!

**Comment by tsauerwein** [comment]
Thanks for the review!

**Pull request merged**